### PR TITLE
[SPARK-24452][SQL][Core] Avoid possible overflow in int add or multiple

### PR DESCRIPTION
**Pull request by kiszk**
## What changes were proposed in this pull request?

This PR fixes possible overflow in int add or multiply. In particular, their overflows in multiply are detected by [Spotbugs](https://spotbugs.github.io/)

The following assignments may cause overflow in right hand side. As a result, the result may be negative.
```
long = int * int
long = int + int
```

To avoid this problem, this PR performs cast from int to long in right hand side.

## How was this patch tested?

Existing UTs.

### CODE REVIEW

**Comment by kiszk** [comment]
cc @ueshin @hvanhovell

**Comment by kiszk** [comment]
cc @cloud-fan

**Comment by JoshRosen** [comment]
Hey @kiszk, thanks for tracking this down. This change looks good to me.

I have a couple of questions, mostly aimed towards figuring out how we can categorically solve this problem:

1. What's the impact of this issue? Have you observed actual crashes or silent data corruption caused by this problem? I ask because it looks like this could be a plausible cause of a data corruption bug that I've been investigating.
2. Are these five files the only occurrence of this problem or are there potentially others? I've noticed that all of the files modified here are `.java` files: is that a coincidence or is that the result of running some Java linting tool? If the latter, is it possible that we have similar issues in other files which we haven't found yet?
3. Do you have any ideas for how we can categorically prevent this class of problem in the future? Are there compiler plugins or linters which can warn on these cases and turn them into compile-time errors? Or code-review checklists that we can employ to more easily spot these potential overflows? This is a hard problem, but I think it's worth brainstorming on categorical solutions since this isn't the first time we've hit this class of problem (but hopefully it can be the last!).

I think we should definitely backport this fix, at least to 2.3 and possibly earlier.

**Comment by cloud-fan** [comment]
+1 on @JoshRosen 's ideas.

I've already seen several overflow fixes from @kiszk , it will be good if we have some tools to check it, even we need to run the tool manually. One idea may be to force to use java 8's safe math functions: https://docs.oracle.com/javase/8/docs/api/java/lang/Math.html#addExact-int-int-

**Comment by kiszk** [comment]
Good questions.

For 2, at first I found one of these issues when I looked at a file. Then, I ran `grep` command with `long .*=.*\*` and `long .*=.*\+` in `.java` file. Then, I picked them up manually. It looks labor-intensive.

For 3, here is my thought.
[`SpotBugs`](https://spotbugs.github.io/) may be a good candidate to check it.
SpotBug is a successor of  [`findBugs`](https://findbugs.sourceforge.net/). When I ran `FindBugs` before, I found some problems regarding possible overflow and then made a PR that was integrated. On the other hand, these issues may not be detected at that time.

I will look at SpotBugs after my presentation at SAIS will be finished :)

**Comment by kiszk** [comment]
@JoshRosen @cloud-fan 
Here is an update.

I have just apply `findBugs` to `OffHeapColumnVector.java` and `UnsafeArrayData.java`.
In `OffHeapColumnVector.java`, most of possible overflows are detected. But, not all.
In `UnsafeArrayData.java`, two possible overflows are detected. Line 86 and 456. I overlooked Line 86.

**Comment by cloud-fan** [comment]
is findBugs available for scala code as well?

**Comment by kiszk** [comment]
Since it is Java bytecode analysis, it is available for Scala code, too.
In my quick test, findBugs overlooked a possible overflow. On the other hand, findBugs found another redundant null check.

**Comment by kiszk** [comment]
Since I found an plug-in for maven, I will also include a patch to add findBugs/SpotBugs into maven in this PR.

**Comment by JoshRosen** [comment]
Let's merge this as-is and do the build improvements in a separate PR. That's important because we may want to backport the overflow fix to maintenance branches and may want to do so independent of the build changes.

**Comment by kiszk** [comment]
Thank you for your comment. I will create another PR for integrating findBugs/SpotBugs into maven.

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/VariableLengthRowBasedKeyValueBatch.java [hunk 1]
```diff
@@ -41,7 +41,7 @@
   @Override
   public UnsafeRow appendRow(Object kbase, long koff, int klen,
                              Object vbase, long voff, int vlen) {
-    final long recordLength = 8 + klen + vlen + 8;
+    final long recordLength = 8L + (long)klen + (long)vlen + 8L;
```
Comment (cloud-fan) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/VariableLengthRowBasedKeyValueBatch.java]: ditto

**Comment by cloud-fan** [comment]
@kiszk  when you were running findBugs locally, did you find more overflow bugs that are not present in this PR? Let's put all discovered overflow bugs in this PR and have another PR to integrate findBugs with maven.

File: core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java [hunk 2]
```diff
@@ -703,7 +703,7 @@ public boolean append(Object kbase, long koff, int klen, Object vbase, long voff
       // must be stored in the same memory page.
       // (8 byte key length) (key) (value) (8 byte pointer to next value)
       int uaoSize = UnsafeAlignedOffset.getUaoSize();
-      final long recordLength = (2 * uaoSize) + klen + vlen + 8;
+      final long recordLength = (2L * uaoSize) + (long)klen + (long)vlen + 8L;
```
[the remark this answers is not in the document]
Comment (kiszk) [reply core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java]: You are right. It was too conservative. ` (2L * uaoSize) + klen + vlen + 8` can generate `LMUL` or `LADD` as follows:
```
    LDC 2
    ILOAD 9
    I2L
    LMUL
    ILOAD 4
    I2L
    LADD
    ILOAD 8
    I2L
    LADD
    LDC 8
    LADD
    LSTORE 10
```

**Comment by kiszk** [comment]
@cloud-fan addressed all of the possible integer overflows detected by SpotBugs.

**Comment by cloud-fan** [comment]
thanks, merging to master/2.3!

**Pull request closed without merging**